### ISSUE TITLE
Project Level 4 

### ISSUE THREAD

**Issue by brianlyc** [nus-cs2030/2021-s2#238]
# Summary
How is the rest time allocated to each server?

## Description 
My understanding of level 4 is that I am supposed to have a counter to count the number of customers that have been served. Then I will always increment this count after each Done event. After a server is done serving a customer, the server will then take this count value and extract the correct rest time from an array of rest time. 
Even though this method worked for the two sample test cases, I still encountered errors for test case 4_3.

## Screenshots (if any): 
![image](https://user-images.githubusercontent.com/78066188/115113478-31519e80-9fbd-11eb-895e-0cadd4464ac6.png)

**Comment by yokester98** [comment]
Hello! I managed to extract out the hidden test case to debug my code. Hope it helps!

2 2 20
0.0 0.313141
0.313508 0.103753
1.204909 0.699522
2.776498 0.014224
3.876961 0.154173
3.909736 0.012659
9.00639 1.477717
9.04336 2.59302
9.105379 0.296847
9.159629 0.368667
9.224613 0.051732
10.147993 3.489595
11.204932 0.006333
12.428914 0.160532
13.109177 2.86915
15.263626 0.576351
15.524295 0.89579
15.939973 1.04302
17.793096 3.007573
18.765423 0.74237
0.0
3.138762
0.847804
0.0
0.848968
0.0
0.0
3.863139
0.0
0.0
25.460873
0.0
0.0
0.0
0.0
36.96365
0.0
3.577445
2.169913
0.0

**Comment by marcuslim835** [comment]
yup you are right in the way you implemented rest time. the error you are getting probably comes from queueing though. the way customers choose a queue should be independent of whether a server is resting

**Comment by gableejh** [comment]
based on the screenshot, looks like the queue for server 1 is full and thus cust 14 should queue at server 2. perhaps you can check the logic for your queueing

**Comment by dumbasssoul** [comment]
> Hello! I managed to extract out the hidden test case to debug my code. Hope it helps!
> 
> 2 2 20
> 0.0 0.313141
> 0.313508 0.103753
> 1.204909 0.699522
> 2.776498 0.014224
> 3.876961 0.154173
> 3.909736 0.012659
> 9.00639 1.477717
> 9.04336 2.59302
> 9.105379 0.296847
> 9.159629 0.368667
> 9.224613 0.051732
> 10.147993 3.489595
> 11.204932 0.006333
> 12.428914 0.160532
> 13.109177 2.86915
> 15.263626 0.576351
> 15.524295 0.89579
> 15.939973 1.04302
> 17.793096 3.007573
> 18.765423 0.74237
> 0.0
> 3.138762
> 0.847804
> 0.0
> 0.848968
> 0.0
> 0.0
> 3.863139
> 0.0
> 0.0
> 25.460873
> 0.0
> 0.0
> 0.0
> 0.0
> 36.96365
> 0.0
> 3.577445
> 2.169913
> 0.0

hello @yokester98 do you have the corrrect output for this?

**Comment by yokester98** [comment]
@dumbasssoul here you go!

0.000 1 arrives
0.000 1 serves by server 1
0.313 1 done serving by server 1
0.314 2 arrives
0.314 2 serves by server 1
0.417 2 done serving by server 1
1.205 3 arrives
1.205 3 serves by server 2
1.904 3 done serving by server 2
2.776 4 arrives
2.776 4 serves by server 2
2.791 4 done serving by server 2
3.877 5 arrives
3.877 5 serves by server 1
3.910 6 arrives
3.910 6 serves by server 2
3.922 6 done serving by server 2
4.031 5 done serving by server 1
9.006 7 arrives
9.006 7 serves by server 1
9.043 8 arrives
9.043 8 serves by server 2
9.105 9 arrives
9.105 9 waits at server 1
9.160 10 arrives
9.160 10 waits at server 1
9.225 11 arrives
9.225 11 waits at server 2
10.148 12 arrives
10.148 12 waits at server 2
10.484 7 done serving by server 1
10.484 9 serves by server 1
10.781 9 done serving by server 1
11.205 13 arrives
11.205 13 waits at server 1
11.636 8 done serving by server 2
11.636 11 serves by server 2
11.688 11 done serving by server 2
11.688 12 serves by server 2
12.429 14 arrives
12.429 14 waits at server 2
13.109 15 arrives
13.109 15 waits at server 2
14.644 10 serves by server 1
15.013 10 done serving by server 1
15.178 12 done serving by server 2
15.178 14 serves by server 2
15.264 16 arrives
15.264 16 waits at server 1
15.338 14 done serving by server 2
15.338 15 serves by server 2
15.524 17 arrives
15.524 17 waits at server 2
15.940 18 arrives
15.940 18 waits at server 2
17.793 19 arrives
17.793 19 leaves
18.207 15 done serving by server 2
18.207 17 serves by server 2
18.765 20 arrives
18.765 20 waits at server 2
19.103 17 done serving by server 2
19.103 18 serves by server 2
20.146 18 done serving by server 2
40.474 13 serves by server 1
40.480 13 done serving by server 1
40.480 16 serves by server 1
41.056 16 done serving by server 1
57.110 20 serves by server 2
57.852 20 done serving by server 2
[6.025 19 1]

**Comment by arjann57** [comment]
hello, i faced a similar issue previously and it was due to my method (in server class) which decides whether a customer can queue at a server. you could test your queuing system by checking whether rest times (and hence next available times of the server) are being kept track of by inserting print statements. in this case, because server 1 has a long rest time, the queue for server 1 is still full and hence the customer should queue at server 2

**Comment by etbf** [comment]
hi i was facing an issue with how a customer gets to queue at a certain server but i dont know if it will help you or not. So basically when i dequeue a next waiting customer in a done event to prompt the server to serve this customer, i shouldnt do it if the server is about to rest as the waiting customer should still be waiting in the queue instead of going up to serve. this has caused my server state to mess up and to debug it i only dequeue the latter case in a ServerWake event that i created. hopefully it somehow helps your code.
